lib/virtio_mmio/virtio_mmio.c

```diff
@@ -423,4 +530,165 @@ void virtio_mmio_isr(struct virtio_device *vdev)
+		if (_cb_funcs[i] != NULL) {
+			_cb_funcs[i](_cb_args[i]);
+		}
+	}
+}
+
+/* weak symbols */
+void virtio_mmio_shm_pool_init(void *mem, size_t size)
+{
+	(void)mem;
+	(void)size;
+}
+
+void *virtio_mmio_shm_pool_alloc(size_t size)
+{
+	(void)size;
+	return NULL;
+}
+
+void virtio_mmio_shm_pool_free(void *ptr)
+{
+	(void)ptr;
+}
+
+void virtio_mmio_hvl_ipi(void) {}
+
+void virtio_mmio_hvl_wait(uint32_t usec)
+{
+	(void)usec;
+}
+
+int virtio_mmio_hvl_wait_cfg(struct virtio_device *vdev, int offset, uint32_t data)
+{
+	uint32_t val;
+
+	val = virtio_mmio_read32(vdev, offset);
+	if (val == data) {
+		return 0;
+	}
+
+	virtio_mmio_hvl_wait(4000000);
+
+	val = virtio_mmio_read32(vdev, offset);
+	if (val == data) {
+		return 0;
+}
+
+	return -1;
+}
+
+
+int virtio_mmio_hvl_wait_cfg_ptr(uint32_t *ptr, uint32_t data)
+{
+	if (*ptr == data) {
+		return 0;
+	}
+
+	virtio_mmio_hvl_wait(4000000);
+
+	if (*ptr == data) {
+		return 0;
+	}
+
+	return -1;
 }
+
+uint64_t virtio_mmio_hvl_add_bounce_buf(struct virtqueue *vq, void *cookie, char *buffer,
+					unsigned int len)
+{
+	struct vq_bounce_buf *bb = NULL;
+	uint64_t addr = 0;
+	char *buf = NULL;
+	struct virtio_mmio_device *vmdev = NULL;
+
+	if (!vq || !cookie || !buffer || (len == 0))
+		return 0;
+
+	vmdev = metal_container_of(vq->vq_dev, struct virtio_mmio_device, vdev);
+
+	if (vmdev && (vmdev->hvl_mode == 1)) {
+		/* check if buffer is already in pre-shared shared memory */
+		addr = metal_io_virt_to_phys(vq->shm_io, buffer);
+		if (addr != METAL_BAD_PHYS) {
+			return addr;
+		}
+
+		buf = virtio_mmio_shm_pool_alloc(len);
+		if (buf != NULL) {
+			memcpy(buf, buffer, len);
+			addr = metal_io_virt_to_phys(vq->shm_io, buf);
+			bb = virtio_mmio_shm_pool_alloc(sizeof(struct vq_bounce_buf));
+			if (bb != NULL) {
+				bb->addr = (uintptr_t)buffer;
+				bb->len = len;
+				bb->cookie = cookie;
+				metal_list_add_tail(&vmdev->bounce_buf_list, &bb->node);
+				return addr;
+			}
+			virtio_mmio_shm_pool_free(buf);
+		}
+	}
+
+	return 0;
+}
+
+void virtio_mmio_hvl_get_bounce_buf(struct virtqueue *vq, struct vring_desc *dp, void *cookie)
+{
+	struct vq_bounce_buf *bb = NULL;
+	struct metal_list *node = NULL;
+	struct virtio_mmio_device *vmdev = NULL;
+
+	VIRTIO_ASSERT(vq != NULL, "Invalid virtqueue");
+	VIRTIO_ASSERT(dp != NULL, "Invalid vring descriptor");
+	VIRTIO_ASSERT(cookie != NULL, "Invalid cookie");
+
+	vmdev = metal_container_of(vq->vq_dev, struct virtio_mmio_device, vdev);
+
+	if (vmdev && (vmdev->hvl_mode == 1)) {
+		metal_list_for_each(&vmdev->bounce_buf_list, node) {
+			bb = metal_container_of(node, struct vq_bounce_buf, node);
+			if (bb->cookie == cookie) {
+				memcpy((void *)(uintptr_t)bb->addr,
+				       (void *)(metal_io_phys_to_virt(vq->shm_io, dp->addr)),
+				       dp->len);
+				metal_list_del(node);
+				virtio_mmio_shm_pool_free(bb);
+				virtio_mmio_shm_pool_free(
+					(void *)(metal_io_phys_to_virt(vq->shm_io, dp->addr)));
+				break;
+			}
+		}
+	}
+}
+
```

Comment:
I don't have a look yet to `virtio_mmio_hvl_add_bounce_buf` and `virtio_mmio_hvl_get_bounce_buf` functions
First, I need to understand the associated use case

Reply:
These routines are the logical equivalents of add and get virtqueue buffers, but they operate on the shared memory buffer pool and handle the memcpy from and to the data buffers used by the virtio front-ends.